unit_tests/test_ops_openstack.py

```diff
@@ -0,0 +1,268 @@
+            'Unit is ready')
+        self.assertIsInstance(
+            self.harness.charm.unit.status,
+            ActiveStatus)
+
+    def test_update_status_custom_check_fail(self):
+        self.harness.update_config(
+            key_values={
+                'custom-check-fail': 'True'})
+        self.harness.add_relation('shared-db', 'mysql')
+        self.harness.begin()
+        self.harness.charm.state.is_started = True
+        self.harness.charm.on_update_status('An Event')
+        self.assertEqual(
+            self.harness.charm.unit.status.message,
+            'Custom check failed')
+        self.assertIsInstance(
+            self.harness.charm.unit.status,
+            MaintenanceStatus)
+
+    def test_update_status_not_started(self):
+        self.harness.add_relation('shared-db', 'mysql')
+        self.harness.begin()
+        self.harness.charm.on_update_status('An Event')
+        self.assertEqual(
+            self.harness.charm.unit.status.message,
+            'Charm configuration in progress')
+        self.assertIsInstance(
+            self.harness.charm.unit.status,
+            WaitingStatus)
+
+    def test_update_status_series_upgrade(self):
+        self.harness.begin()
+        self.harness.charm.state.series_upgrade = True
+        self.harness.charm.on_update_status('An Event')
+        self.assertEqual(
+            self.harness.charm.unit.status.message,
+            ('Ready for do-release-upgrade and reboot. Set complete when '
+             'finished.'))
+        self.assertIsInstance(
+            self.harness.charm.unit.status,
+            BlockedStatus)
+
+    def test_update_status_series_paused(self):
+        self.harness.begin()
+        self.harness.charm.state.is_paused = True
+        self.harness.charm.on_update_status('An Event')
+        self.assertEqual(
+            self.harness.charm.unit.status.message,
+            "Paused. Use 'resume' action to resume normal service.")
+        self.assertIsInstance(
+            self.harness.charm.unit.status,
+            MaintenanceStatus)
+
+    def test_update_status_missing_relation(self):
+        self.harness.begin()
+        self.harness.charm.on_update_status('An Event')
+        self.assertEqual(
+            self.harness.charm.unit.status.message,
+            'Missing relations: shared-db')
+        self.assertIsInstance(
+            self.harness.charm.unit.status,
+            BlockedStatus)
+
+    def test_services(self):
+        self.harness.begin()
+        self.assertEqual(
+            self.harness.charm.services(),
+            ['apache2', 'ks-api'])
+
+    def test_pre_series_upgrade(self):
+        self.os_utils.manage_payload_services.return_value = ('a', 'b')
+        self.harness.begin()
+        self.assertFalse(self.harness.charm.state.series_upgrade)
+        self.assertFalse(self.harness.charm.state.is_paused)
+        self.harness.charm.on_pre_series_upgrade('An Event')
```

Comment:
Similar to the comment about update-status, could emit this event instead:

`self.harness.charm.on.pre_series_upgrade.emit()`